### ISSUE TITLE
Better user documentation

### ISSUE THREAD

**Issue by jcbhmr** [devcontainers-contrib/features#69]
Right now things don't look to great when you wind up on the repo home page.

![](https://thum.io/get/https://github.com/devcontainers-contrib/features/tree/a87e6d2257d62c327d9edc0bbe8be1746c701815)

To change this, I think that these things could be done:
- [x] Add a header image #68
- [x] Add a more descriptive description #68 
- [x] Add a "how to use" section #68 
- [x] Add a CONTRIBUTING.md guide #68 
- [x] ~~Add details on how to add a new feature~~ Moved to another "better contributor docs" issue #77

**Comment by jcbhmr** [comment]
I humbly suggest that one of the admins with the power to do change the repository's "Description" field from "Contributions are Welcomed!" to "🐳 Extra add-in features for Devcontainers and GitHub Codespaces". This description field is sort of the "title" for Google search results. I think it would be wise to make it a more substantive description.

![image](https://user-images.githubusercontent.com/61068799/203651289-40077e88-8b7f-4b07-9076-d45e0bcd9996.png)

**Comment by jcbhmr** [comment]
My honest opinion on the whole Cookiecutter thing:

**It's too complicated.** There is too much logic concentrated in a template file that belongs in separate scripting files. Cookiecutter and Yeoman (and any project generator really) are great tools, but I don't think they fit the bill the best here.

They _work_ but it's unwieldy to edit them. The files are too big for what they should be. There should be a separate `npm-devcontainer-feature.json` and one for `pip-devcontainer-feature.json` if you keep Cookiecutter IMO. They also spread the logic too far apart. You need to reference variables that you declare in JSON and use in Jinja. You need to declare functions separately. You need dependency management. It's a mess.

An alternative is to split each "type" (NPM, pip, apt) into its own Cookiecutter package, but I think a better option is to ditch Cookiecutter as too overkill for simple file generation.

Take Angular for instance. They use [`ng generate <type>`](https://angular.io/cli/generate) as a way to create new components, interfaces, app shells, and more. I think this approach of script-based generation should be adopted going forward instead of more complex unweidly (and frankly very intimidating) Cookiecutter template logic. I mean _have you seen [that `devcontainer-feature.json` file](https://github.com/devcontainers-contrib/features/blob/main/pkgs/feature-template/%7B%7Bcookiecutter.base_dir%7D%7D/src/%7B%7Bcookiecutter.feature_id%7D%7D/devcontainer-feature.json)?_ It is write-once and never touch again code at its finest.

Another argument is that _we really need freaking [**examples** of how to do this](https://github.com/devcontainers-contrib/features/tree/main/pkgs/feature-template/examples)?_ Is it really that complicated to just _fill out a form_ in your terminal basically?

Here's an example to really nail home what I am talking about. This is a complete Feature generator script that **works**! I just wrote it so it might not have proper error handling. It generates an NPM-based feature.

Try it out on Replit! It actually works! https://replit.com/@jcbhmr/SquigglyRepulsiveSolaris#index.js

```js
#!/usr/bin/env -S deno run -A
const npmName = prompt("What is the name of the NPM package?");
const featName = prompt("What should the feature name be?");
const featId = prompt("What should the feature ID be?");
const featDesc = prompt("What should the feature description be?");
const featTestCmd = prompt("What should the test command be?");

const featObject = {
  name: featName,
  description: featDesc,
  id: featId,
  version: "1.0.0",
  documentationURL: `https://github.com/devcontainers-contrib/features/tree/main/src/${featId}#readme`,
  installAfter: ["ghcr.io/devcontainers/features/node:1"],
  options: {
    version: {
      type: "boolean",
      proposals: ["latest"],
      default: "latest",
      description: `Select the version of ${npmName}`,
    },
  },
};
const featJSON = JSON.stringify(featObject, 0, 4);

const featInstallSH = `#!/usr/bin/env bash
set -Eeuo pipefail

if ! command -v npm &>/dev/null; then
  echo "WARN: 'npm' was not found. Installing Node.js + NPM using devcontainer/features src/node/install.sh..."
  curl -fsSL https://raw.githubusercontent.com/devcontainers/features/main/src/node/install.sh | /bin/bash
fi

npm i -g ${npmName}@\${VERSION:-"latest"}
`;

const featTestSH = `#!/usr/bin/env bash
set -Eeuo pipefail

${featTestCmd}
`;

console.log(`src/${featId}/devcontainer-feature.json`);
console.log(featJSON);
console.log(`src/${featId}/install.sh`);
console.log(featInstallSH);
console.log(`test/${featId}/test.sh`);
console.log(featTestSH);

await Deno.mkdir(`src/${featId}`, { recursive: true }); // Same as 'mkdir -p <folder>'.
await Deno.writeTextFile(`src/${featId}/devcontainer-feature.json`, featJSON);
await Deno.writeTextFile(`src/${featId}/install.sh`, featInstallSH);
await Deno.chmod(`src/${featId}/install.sh`, 0o755); // Default of 'chmod +x <file>' is 755.
await Deno.mkdir(`test/${featId}`, { recursive: true });
await Deno.writeTextFile(`test/${featId}/test.sh`, featTestSH);
```

![image](https://user-images.githubusercontent.com/61068799/203683471-3803cb89-95f4-43ab-b33a-dcec13a2ae6a.png)

Then the user experience is so much better! You just run the script and it does it for you. You can even "infer" things from the current Git repo like the remote origin URL for documentation URL generation! And it makes it easier to add new features without coupling them to other features (ie. not as many if-statements). **I know you can probably do that in Cookiecutter too, but this is _one file_ instead of like 5 files in 3 directories. It's easier to reason about, follow, and maintain.** Especially when compared to that monster of an `install.sh` template.

For instance, here are some generator script ideas:
- `scripts/generate-new-feature-pip.js`
- `scripts/generate-new-feature-apt.py`
- `scripts/generate-new-feature-deno.sh`
- `scripts/generate-new-feature-brew.rb`
- `scripts/generate-new-feature-${yourPackageManager}.${yourFavoriteScriptingLanguage}`

Heck, you could even commit [a C++ file that compiled and ran itself](https://gist.github.com/jdarpinian/1952a58b823222627cc1a8b83a7aa4e2) if you really wanted to!

**TL;DR:** I want a 👍✅ or a 👎⛔ before I invest more time. If I do all the work and the PR doesn't merge, that's not fun for me 😭.

---

PS: Yes, you can use Python if you're into Python over TypeScript/JavaScript. I personally like the ability of JavaScript + Deno to be self-contained in a single file _even if it includes library stuff_ using HTTP import URLs.

```js
// Using the Prettier code formatter which is relatively large. It's one URL import away!
// https://github.com/denolib/prettier#readme
import { prettier, prettierPlugins } from "https://denolib.com/denolib/prettier/prettier.ts";

const json = JSON.stringify({ my: "object" })
const opts = { parser: "json", plugins: prettierPlugins };
prettier.format(json, opts);
//=> '{ "my": "object" }\n'
```
